Improve responsiveness of a wxGrid having plenty of attributes

discnl opened the pull request:
Speed up grid attribute lookups by using a hash map instead of array.

Closes [#12764](https://trac.wxwidgets.org/ticket/12764).

--

Please see the referenced ticket for further details.

There's another optimization possible, before and still with this PR, by not always doing two lookups when getting a new attr in [wxGrid::GetOrCreateCellAttr](https://github.com/wxWidgets/wxWidgets/blob/36ab71301c5cd4390881ea901bbce7d4da6e5f9f/src/generic/grid.cpp#L9387-L9394). The first lookup is with `GetAttr()` then when that returns `NULL` there's another lookup through `SetAttr()`. It could be changed by having something like `SetNewAttr()` (twice, for `wxGridTableBase` and `wxGridCellAttrProvider`) and `wxGridCellAttrData::SetAttr()` getting a bool parameter `isConfirmedNewAttr` to skip lookup if true. However I don't know if it's worth the hassle because remaining backwards compatible would seem to require more changes, as there needs to be a way to know if a derived class handles `SetNewAttr()`. Probably by having a `virtual bool DoSetNewAttr()` (called by non-virtual `SetNewAttr()`), and calling `SetAttr()` like before if it returns false (not handled).

discnl committed:
Demonstrate intensive attribute usage in grid sample

Add options to fill grid either half (checkered cells) or fully (all
cells coloured) with attributes. Available from Grid > "Toggle Checkered
Cells" and "Toggle Coloured Cells".

discnl committed:
Make slowness of attribute lookups more apparent in grid sample

Multiply the number of rows in the grid sample by a factor of 10. This
demonstrates the slowness with attribute lookups better when using
checkered (Ctrl+Shift+K) or coloured (Ctrl+Shift+L) cells, as filling
the grid is slowed down by way more than a factor of 10. Scrolling also
becomes less responsive the further away from the home cell the view is.

As a result of increasing the row count also move the cells positioned
near the end of the grid there again, and don't use fixed coords for
them.

discnl committed:
Improve responsiveness of a wxGrid having plenty of attributes

Speed up grid attribute lookups by using a hash map instead of array.

Closes #12764.

discnl committed:
Remove wxGridCellWithAttr

After the previous commit the coords stored by wxGridCellWithAttr have
become redundant as the coords are now also stored as part of the key
used for an attributes map. With then only an attribute remaining in
wxGridCellWithAttr it can be removed completely, letting the value of
an attributes map point directly to a (ref-counted) attribute instead.

vadz reviewed (approved):
Thanks! This definitely looks good to me and I like that we can get rid of the confusing `wxGridCellWithAttr` class completely.

Just to make sure: I think this code is already exercised by the existing grid tests, but could you please confirm this?

Also, there are a couple of nitpicks in the sample which don't prevent this from being merged, but it would be nice to fix them (especially as it should take just a couple of minutes). TIA!

discnl commented:
> Thanks! This definitely looks good to me and I like that we can get rid of the confusing `wxGridCellWithAttr` class completely.

That's good to hear, as I thought its removal may actually have been disputable for some reason. If no objection I'll squash the last two commits then that are about removing `wxGridCellWithAttr`. The last three could be squashed but in 204e38ae15 there's an assert failure if the coords stored in the key of an attribute map are different from the coords in `wxGridCellWithAttr` that may be valuable to retain (even if just to show that their synched states at least were checked at some point).

> Just to make sure: I think this code is already exercised by the existing grid tests, but could you please confirm this?

I reviewed again and I'm actually not confident about the overwriting of an attribute (where `wxGridCellWithAttr`'s `ChangeAttr()` used to be called), for the tests anyway as the grid sample does call that code. I think the earlier added basic attribute tests cover it too though, will verify later.

Regarding reliability of `UpdateCellAttrRowsOrCols` changes: (all of) this code existed before master's multicell fixes with an array of attributes, so master changes were derived from these using a hash map. The newer multicell tests also should cover it completely (will do a poor man's impression of code coverage by sprinkling breakpoints here as well to verify).

> Also, there are a couple of nitpicks in the sample which don't prevent this from being merged, but it would be nice to fix them (especially as it should take just a couple of minutes). TIA!

Thanks, I will process them along with the other checking later on.

discnl commented on samples/grid/griddemo.cpp:
```diff
@@ -2850,3 +2859,86 @@ void GridFrame::ShowRow( wxCommandEvent& WXUNUSED(event) )
+// colour changes.
+void ToggleGridCells(wxGrid* grid, bool useCheckered)
+{
+    static bool _apply[2] = { };
+    const bool apply = _apply[useCheckered] = !_apply[useCheckered];
```
Famous last-minute changes, I must have been blinded by introducing `apply` and not repeating `apply[useCheckered]` for a whole two times or something. Just before it was:

```cpp
static bool apply[2] = { };
apply[useCheckered] = !apply[useCheckered];
```

And what if `useCheckered` (/`effectIndex`) is `unsigned` (or rather `unsigned int` in wx) instead?

Then again I guess two statics could just be used, I just felt like it could discourage the addition of, for whatever reason, another effect somewhat.

I suppose the single underscore avoiding is a style thing (it's reserved for globals was my conclusion when checking). I sometimes use the ugly underscore exactly to discourage using such variables further in code, as here convenience variable `apply` should only have been used instead.

discnl committed:
Add minor test for grid attributes

Test overwriting a cell's attribute with another attribute. With this
addition wxGridCellAttrData::Setattr() is now fully covered by tests.

discnl commented:
There was one path that got changed and wasn't already covered by the grid tests, I prepended b5edfa2872 for that. It's a minor test but there's not much more to test with it I suppose. There's such a thing as overlapping merged cells but right now I'd rather not go there regarding expected behaviour or even existence.
Also verified that `UpdateCellAttrRowsOrCols()` is fully covered by tests, made review changes (thanks!), squashed and rebased on master.

vadz commented:
Thanks! I'll have one final look at this slightly later, for now I just wanted to mention that the CI builds are due to spurious failures again.

The AppVeyor one failed with
```
-------------------------------------------------------------------------------
WebRequest::SSL::Ignore
-------------------------------------------------------------------------------
./net/webrequest.cpp:288
...............................................................................
./net/webrequest.cpp:141: 
warning:
  Error: The buffers supplied to a function was too small.
./net/webrequest.cpp:144: FAILED:
  REQUIRE( request.GetState() == requiredState )
with expansion:
  4 == 3
```
which looks like a real bug in our code (@MaartenBent please let me know if you're going to look into this, by chance, so that I don't duplicate your efforts), while Travis CI fails when installing httpbin from Docker with
```
anyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```
and the solution here is probably just to stop using Docker and use pip (so much for "it's more robust" comment...).

vadz commented:
Thanks again for all your work on this!
